binr/r2r/r2r.c

```diff
@@ -532,15 +534,231 @@ static RThreadFunctionRet worker_th(RThread *th) {
+						st16 tl_score = *(align_table + (row - 1) * dim + col - 1)
+								+ (expected[col - 1] == actual[row - 1] ? match : mismatch);
+						st16 t_score = *(align_table + (row - 1) * dim + col) + gap;
+						st16 l_score = *(align_table + row * dim + col - 1) + gap;
+						st16 score;
+						if (tl_score >= t_score && tl_score >= l_score) {
+							score = tl_score;
+						} else if (t_score >= tl_score && t_score >= l_score) {
+							score = t_score;
+						} else {
+							score = l_score;
+						}
+						*(align_table + row * dim + col) = score;
+					}
+				}
+				// Print table (Debug)
+				char char_str[3] = { ' ' };
+				printf ("%4s ", char_str);
+				for (col = 0; col < dim; col++) {
+					if (col && expected[col - 1] == '\n') {
+						char_str[0] = '\\';
+						char_str[1] = 'n';
+					} else {
+						char_str[0] = !col ? ' ' : expected[col - 1];
+						char_str[1] = 0;
+					}
+					printf ("%4s ", char_str);
+				}
+				printf ("\n");
+				for (row = 0; row < dim; row++) {
+					if (row && actual[row - 1] == '\n') {
+						char_str[0] = '\\';
+						char_str[1] = 'n';
+					} else {
+						char_str[0] = !row ? ' ' : actual[row - 1];
+						char_str[1] = 0;
+					}
+					printf ("%4s ", char_str);
+					for (col = 0; col < dim; col++) {
+						printf ("%4d ", *(align_table + row * dim + col));
+					}
+					printf ("\n");
+				}
+				// Do alignment
+				size_t idx_expected = len_expected - 1;
+				size_t idx_actual = len_actual - 1;
+				size_t idx_align = 2 * len_expected - 1;
+				size_t pos_row = dim - 1;
+				size_t pos_col = dim - 1;
+				while (pos_row != 0 || pos_col != 0) {
+					st16 tl_score = ST16_MIN;
+					st16 t_score = ST16_MIN;
+					st16 l_score = ST16_MIN;
+					if (pos_row > 0) {
+						t_score = *(align_table + (pos_row - 1) * dim + pos_col);
+					}
+					if (pos_col > 0) {
+						l_score = *(align_table + pos_row * dim + pos_col - 1);
+					}
+					if (pos_row > 0 && pos_col > 0) {
+						tl_score = *(align_table + (pos_row - 1) * dim + pos_col - 1);
+					}
+					if (t_score >= tl_score && t_score >= l_score) {
+						align_expected[idx_align] = 0;
+						align_actual[idx_align] = actual[idx_actual--];
+						idx_align--;
+						pos_row--;
+					} else if (l_score >= tl_score && l_score >= t_score) {
+						align_expected[idx_align] = expected[idx_expected--];
+						align_actual[idx_align] = 0;
+						idx_align--;
+						pos_col--;
+					} else {
+						align_expected[idx_align] = expected[idx_expected--];
+						align_actual[idx_align] = actual[idx_actual--];
+						idx_align--;
+						pos_row--;
+						pos_col--;
+					}
+				}
+				idx_align++;
+				size_t start_align = idx_align;
+				// TODO Print alignment (Debug)
+				for (; idx_align < 2 * len_expected; idx_align++) {
+					ut8 ch = align_expected[idx_align];
+					if (align_actual[idx_align] == '\n' && ch != '\n') {
+						printf (ch ? " " : "-");
+					}
+					if (ch == 0) {
+						printf ("-");
+					} else if (ch == '\n') {
```

Comment:
what about \r and other non printable chars? like > 127?

Reply:
I plan to `#if 0` the code marked `(Debug)`, but to answer the questions:

> what about \r and other non printable chars?

Not sure yet, will probably see how `git diff` handles them.

> like > 127?

Multibyte characters need to be handled in a special manner. Haven't figured out the details.

Right now this pr has been going on for too long and I'm only targeting Minimum Viable Product here, which means ascii-only printable diffs which I guess should suffice for 90-95% of r2r use cases.

Reply:
"Multibyte characters" -> "Multibyte UTF8 characters"